provider/resource_wireguard_preshared_key.go

```diff
@@ -0,0 +1,55 @@
+package provider
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	"io"
+)
+
+func resourceWireguardPresharedKey() *schema.Resource {
+	return &schema.Resource{
+		Description: "Provides a WireGuard asymmetric key resource. This can be used to create, read, and delete WireGuard keys in terraform state.",
+
+		Create: resourceWireguardPresharedKeyCreate,
+		Read:   resourceWireguardPresharedKeyRead,
+		Delete: resourceWireguardPresharedKeyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"preshared_key": {
+				Description: "Additional layer of symmetric-key cryptography to be mixed into the already existing public-key cryptography, for post-quantum resistance.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+		},
+	}
+}
+
+func resourceWireguardPresharedKeyCreate(d *schema.ResourceData, m interface{}) error {
+	var key wgtypes.Key
+	var err error
+
+	key, err = wgtypes.GenerateKey()
+	err = d.Set("preshared_key", key.String())
+	if err != nil {
+		return err
+	}
+	h := md5.New()
+	_, err = io.WriteString(h, key.String())
```

Comment:
This gave me pause at first, but I think it's fine since Terraform's `sensitive` isn't really designed to make anything more secure anyway, it's only avoiding displaying/logging secrets.

Perhaps we should use `crypto/sha256` or `sha512` instead of `md5` though.

Reply:
As you prefer.